enhancements/machine-config/machine-config-node.md

```diff
@@ -0,0 +1,340 @@
+    Last Transition Time:  2023-10-17T13:05:02Z
+    Message:               MachineConfigs Differ, shutting down config drift monitor and comparing the MCs
+    Reason:                NodeUpgradePreparingComparingMCs
+    Status:                False
+    Type:                  ComparingMCs
+	Last Transition Time:  2023-10-17T13:06:02Z
+    Message:               Draining node. Desired Drainer is drain-rendered-master-03c160a7e646004356a0b038dd22a97a
+    Reason:                NodeUpgradeInProgressDraining
+    Status:                False
+    Type:                  DrainingNode
+	Last Transition Time:  2023-10-17T13:06:32Z
+    Message:               Applying files to node. Using new Ignition files:...
+    Reason:                UpgradeInProgressApplyingFilesAndOSChanges
+    Status:                False
+    Type:                  ApplyingFilesAndOSChanges
+	Last Transition Time:  2023-10-17T13:06:40Z
+    Message:               Node will reboot into config rendered-master-03c160a7e646004356a0b038dd22a97a
+    Reason:                UpgradePostActionRebootingNode
+    Status:                False
+    Type:                  RebootingNode
+```
+
+There are two levels of conditions in the MachineConfigNode type: the parent and the child condition. Parent conditions incude Updated, UpdatePreparing, UpdateInProgress, UpdatePostAction, UpdateCompleting, and Resuming. These parent conditions track the overall arc of an upgrade. However, there are often multiple phases within these overarching ones. Therefore: ComparingMCs, DrainingNode, ApplyingFilesAndOS, CordoningNode, RebootingNode, and ReloadingCRIO are the ChildrenPhases that occur during the larger ones. These don't always occur depending on the type of upgrade but sometimes they all occur, leading to some confusion in the parent phases as to what happened. 
+
+
+### API Extensions
+
+TBD
+
+### Risks and Mitigations
+
+There might be users who do not know to look here. We will mititagte this by reporting in the CO and/or the node status go to
+and look at this object for MCO related node failures or progression.
+
+Adding new MCD node states might have unintended consequences on the current flow which just has "Working" and "Done". Will make sure the
+states I add are either only in between Working and Done OR no matter what we end up back in the "Done" state.
+
+### Drawbacks
+
+TBD
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+TBD
+
+### Version Skew Strategy
+
+### Operational Aspects of API Extensions
+
+#### Failure Modes
+
+#### Support Procedures
+
+## Implementation History
+
+## Alternatives
+
+Implementation wise, there has been discussion of the MCD owning these MachineConfigNode objects. While this would work, it contradicts the goal of the MCO's current and future architecture.
```

Comment:
If I recall correctly the concern I was thinking of is reduced RBAC in the daemon. Now this in particular is more of a final state reporting and not really anything dangerous, but from historical perspective, the less we have access to on the daemon (which will run on workers) the better.

To take a step back, if I understand correctly, your concern here is that the current approach introduces data flow like this:
`daemon -> node -> controller -> MCState`
as opposed to a more simplified
`daemon -> node + MCState`

Which I think is a fair concern. I think for us on the MCO team, centralizing state and metrics was the goal we decided upon early, and we're designing around that goal for the current architecture. In this perspective we're increasing complexity of data flow to decrease complexity of decentralized management. The benefits and drawbacks I think aren't hard blockers in either way, but at least I feel like:
1. this allows for metrics to be simplified, so not each node will have to run it's own MCO metrics server
2. in larger scales (1000s of nodes) we don't have 1000 processes trying to mess with machinestates, instead we have just the controller
3. nodes that are down and not running the daemon still has something updating it's state (although most of the time that will just mean no status updates until you fix it, so maybe not really a large benefit)

Which I feel for the MCO team justifies the slight decrease in data efficiency. But we can always look to adjust our design depending on which parts of it you feel like do not justify the goals we set out.

> Are the node annotations present today sufficiently expressive for you to know exactly what the MCD is doing (ie which phase and what errors came out of it)

Sort of, the MCD status and states today are much more vague than what we're looking to do here.

> How do you prevent misleading information from being put into these objects?

The status reporting is just reporting on what it can see. You can certainly modify the node object today to trick the MCD into doing something specific to troubleshoot. In the way the MCO architecture is today, I don't think anything that can trick the state controller would be any worse than the way we can trick the MCO as it stands

Reply:
I also wanted to add a potential compromise here to alleviate worry. I could make it so that the MCD updates the spec of the MCN (which is the desired config basically). Since the MSC only updates the status of the object when the desired config in the spec changes, we could lock it down in that way and only let the daemon modify the spec?

Not sure if that is possible to only allow certain people to modify spec fields, or this is something we could enforce when the MCO goes to update the CRs. but overall what jerry said above is true. At the end of the day the goal here is to decrease the complexity of our current system and that sort of necessitates this design, not just putting more weight on the daemon.  the MCO team has invested a lot of time into the state controller for 4.15 and I believe it will solve a lot of our internal and customer facing issues :)

Reply:
could use something like this: https://github.com/openshift/cluster-node-tuning-operator/blob/7ad3f9bd423334337395d4d0741c3b04a68199a0/pkg/apis/performanceprofile/v2/performanceprofile_validation.go#L50 (seems to be done all over openshift) to validate updates of API types. I could check if the desired config put into the MCN spec is 1) legitimate and 2) matches the desired config on node. Or somehow validate the fieldManager to make sure the MCO was the one who updated this object?

Reply:
I think at the end of the day, the decision on the internal mechanics of this, is up to the MCO team as you'll be the ones maintaining it. In general, when questions are raised on EPs about the mechanics, its convention to document the answers to those, so I would expect an action from this thread to be to document the questions and the responses in a way that explains why the team is not taking on board the suggestion.

Now to counter some of the above arguments, for your consideration.

On the RBAC front, you could create a binding that allows the Node group (`system:nodes`) to update the `MachineConfigNode`, then, with the new [CEL Validating Admission Policy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/), you could create a validation such that, if the user requesting the update is the kubelet credential (as you already do, this is always `system:node:<nodename>`), you can validate that the MCN name being updated matches the `<nodename>` part of the user's name.
This is the same level of validation that the NodeAdmission gives for kubelet today when updating the Node object, so your RBAC would be equivalent to the Kubelet's RBAC for nodes, and, it's pretty simple to implement these admission policies.

> To take a step back, if I understand correctly, your concern here is that the current approach introduces data flow like this:
daemon -> node -> controller -> MCState
as opposed to a more simplified
daemon -> node + MCState

I'm concerned on a couple of fronts with this flow, firstly, there's the [confused deputy](https://docs.aws.amazon.com/IAM/latest/UserGuide/confused-deputy.html) problem, this flow exposes you to a potential privilege escalation. Since nothing is taking action based on the MCN (yet?) this isn't necessarily a major issue right now, but it certainly would be if we ever made the MCN the source of truth for any data. Setting this problem up now will make it harder (you'll want to less) to fix this down the line.

The second issue, is the limitations of the available data. If the MCD updates the MCN directly, it has the opportunity to be more verbose/accurate with the state reporting. With the current flow, it's observing and aggregating existing data, inferring what is happening from a limited source.

There's ongoing conversation about the conditions and how we report the status of the Node in the MCN, but, right now, I don't see how you know which phase the MCD is in, AFAIK that's not on the node annotations, yet, the MCD knows that very well, and could easily update the MCN status to reflect where it is in its processing. Personally, I think you're losing the simplicity of just doing this at the source and that an aggregation is adding complexity, not taking it away.

> Sort of, the MCD status and states today are much more vague than what we're looking to do here.

If you decide later that you want more detail in the MCN, I expect you'll then ask to put new detail onto the Node annotations to accommodate the improved reporting, which would exacerbate this problem IMO and, for the record, not pass an API review.

> in larger scales (1000s of nodes) we don't have 1000 processes trying to mess with machinestates, instead we have just the controller

Have you concerned the reactiveness of distributed workloads vs a single controller? 1 controller processing 1000s of events will be hot and potentially slow to react, where each MCD updating its own will add a small load to each node, but be more reactive in general. Again I think we have prior art here, note that the kubelet on each node, updates the Node object

> this allows for metrics to be simplified, so not each node will have to run it's own MCO metrics server

IIUC this is now a separate enhancement?

> nodes that are down and not running the daemon still has something updating it's state (although most of the time that will just mean no status updates until you fix it, so maybe not really a large benefit)

That's a good counter to my points, but, how often will the state controller be able to update the object if the MCD is down? The MCD is the source of truth for the data, so, if the source is not updating the Node, I don't see how the MCN can be changed?
Any update that can be performed, should probably be performed in cluster anyway, remember there's no reason you have to have strictly just the MCD or just a controller updating the status of the object if there are multiple sources of data

> The status reporting is just reporting on what it can see. You can certainly modify the node object today to trick the MCD into doing something specific to troubleshoot. In the way the MCO architecture is today, I don't think anything that can trick the state controller would be any worse than the way we can trick the MCO as it stands

I think this backs up my data flow concerns. The Node object annotations don't reflect the entire state, but the MCD knows the state better. And there's already a potential issue with your API in that you can confuse the MCD already 😅

> I could make it so that the MCD updates the spec of the MCN (which is the desired config basically).

I'm more concerned about the status of the obejct. The spec here is minimal and should be updated by the MCP(?) when it wants the desired config to change (well, whatever it is that sets that desired). My concerns here are about making sure we have the data available to build a comprehensive and accurate status. 

> Not sure if that is possible to only allow certain people to modify spec fields

This is possible with the validation admission policy stuff I mentioned earlier if that's something you decide you need at some point.

> At the end of the day the goal here is to decrease the complexity of our current system and that sort of necessitates this design, not just putting more weight on the daemon.

Have you looked at what it would take to update a condition using SSA inside the various phases of the MCD, vs the controller based approach? I suspect you could actually make the MCD approach pretty lightweight by giving a different manager to each of your condition types.